File: src/cdm_wizard/wizard.py
import json
import sys
import questionary
from rich.console import Console
from rich.table import Table
from rich.panel import Panel

from cdm_wizard.model import (
    ASSET_DESCRIPTIONS,
    DATA_FIELDS,
    FUNCTION_DESCRIPTIONS,
    GOVERN_FUNCTIONS,
    RIGHT_OF_BOOM_FUNCTIONS,
    TECH_EXAMPLES,
    asset_classes,
    empty_matrix,
    functions,
)

console = Console()

def load_from_json(path: str) -> dict:
    """
    Load CDM data from a JSON file.

    Expected structure (unknown assets/functions are ignored; missing fields
    default to empty strings):

        {
          "Devices": {
            "Identify": {"Tech": "CrowdStrike", "People": "SysAdmin", "Process": "Patch SOP"},
            ...
          },
          ...
        }

    Only the canonical asset classes and NIST functions are accepted; any extra
    keys in the JSON are silently dropped so test fixtures can include comments
    or metadata under non-CDM keys without breaking the export.
    """
    assets = asset_classes()
    function_list = functions()

    with open(path) as f:
        raw = json.load(f)

    if not isinstance(raw, dict):
        raise ValueError(f"JSON root must be an object, got {type(raw).__name__}")

    # Build a clean data structure with defaults, accepting only known keys.
    data = empty_matrix()

    unknown_assets = [k for k in raw if k not in assets]
    if unknown_assets:
        console.print(f"[yellow]Warning:[/yellow] Ignoring unknown asset keys: {unknown_assets}")

    for asset in assets:
        if asset not in raw:
            continue
        asset_data = raw[asset]
        if not isinstance(asset_data, dict):
            raise ValueError(f"Asset '{asset}' must be an object, got {type(asset_data).__name__}")

        unknown_funcs = [k for k in asset_data if k not in function_list]
        if unknown_funcs:
            console.print(f"[yellow]Warning:[/yellow] {asset}: ignoring unknown function keys: {unknown_funcs}")

        for func in function_list:
            if func not in asset_data:
                continue
            cell = asset_data[func]
            if not isinstance(cell, dict):
                raise ValueError(f"Cell '{asset}.{func}' must be an object, got {type(cell).__name__}")
            data[asset][func] = {field: str(cell.get(field, "")) for field in DATA_FIELDS}

    return data


def save_to_json(data: dict, path: str) -> None:
    with open(path, "w") as f:
        json.dump(data, f, indent=2)


def run_wizard():
    """
    Runs the interactive CLI wizard to collect CDM data.
    """
    assets = asset_classes()
    # Move Govern functions to the end for the wizard sequence
    all_functions = functions()
    function_list = [f for f in all_functions if f not in GOVERN_FUNCTIONS] + [f for f in all_functions if f in GOVERN_FUNCTIONS]
    data = empty_matrix()

    console.print(Panel(
        "[bold blue]Cyber Defense Matrix Wizard[/bold blue]\n"
        "[italic]Guided by Sounil Yu's 'The Essential Guide to Navigating the Cybersecurity Landscape'[/italic]\n\n"
        "The CDM is a [bold]Mutually Exclusive and Collectively Exhaustive (MECE)[/bold] framework \n"
        "designed to provide structural clarity and identify gaps in your security posture."
    ))

    console.print(
        "\n[bold yellow]Core Concepts:[/bold yellow]\n"
        "• [bold blue]Left of Boom (Proactive):[/bold blue] [bold]Identify[/bold] and [bold]Protect[/bold]. Focuses on [italic]Structural Awareness[/italic]\n"
        "  (inventory, vulnerability discovery, and preventative safeguards).\n"
        "• [bold red]Right of Boom (Reactive):[/bold red] [bold]Detect[/bold], [bold]Respond[/bold], and [bold]Recover[/bold]. Focuses on [italic]Situational Awareness[/italic]\n"
        "  (identifying exploitations, containment, and restoration).\n"
        "\n[bold yellow]The Dependency Continuum:[/bold yellow]\n"
        "• [bold]Technology[/bold] dependency is strongest in [blue]Identify/Protect[/blue].\n"
        "• [bold]People[/bold] dependency grows significantly in [red]Detect/Respond/Recover[/red].\n"
        "• [bold]Process[/bold] is the essential, consistent foundation throughout all functions.\n"
    )

    from cdm_wizard.model import ASSET_ICONS

    total_cells = len(assets) * len(function_list)

    for asset_idx, asset in enumerate(assets):
        icon = ASSET_ICONS.get(asset, "🔹")
        console.print(f"\n[bold cyan]{icon} {asset.upper()} " + ("━" * (console.width - len(asset) - 8)) + "[/bold cyan]")
        console.print(f"  [italic dim cyan]{ASSET_DESCRIPTIONS[asset]}[/italic dim cyan]")

        for func_idx, func in enumerate(function_list):
            cell_num = asset_idx * len(function_list) + func_idx + 1
            progress = f"[dim](Asset {asset_idx + 1}/{len(assets)} | Function {func_idx + 1}/{len(function_list)} | Cell {cell_num}/{total_cells})[/dim]"

            if func in GOVERN_FUNCTIONS:
                boom_tag = "[bold yellow]Cross-cutting[/bold yellow]"
                people_instruction = "e.g. Security Governance Lead, Risk Committee"
                process_instruction = "e.g. Risk Exception Process, Security Policy Review"
            else:
                is_right_of_boom = func in RIGHT_OF_BOOM_FUNCTIONS
                boom_tag = "[bold red]Right of Boom[/bold red]" if is_right_of_boom else "[bold blue]Left of Boom[/bold blue]"
                people_instruction = "e.g. SOC Analyst, SysAdmin"
                process_instruction = "e.g. Patch Management SOP, IR Plan"

            tech_instruction = TECH_EXAMPLES.get((asset, func), "e.g. vendor tool or platform name")

            console.print(f"\n  [bold reverse] {func} [/bold reverse] [bold]{boom_tag}[/bold] {progress}")
            console.print(f"  [dim]{FUNCTION_DESCRIPTIONS[func]}[/dim]")

            tip_prefix = "  [italic yellow]💡 Tip:[/italic yellow]"
            if func == "Govern":
                console.print(f"{tip_prefix} [italic dim]Think of this as the 'Context'—policies, risk appetite, and oversight governing this asset class.[/italic dim]")
            elif func in ("Identify", "Protect", "Recover"):
                console.print(f"{tip_prefix} [italic dim]Map based on the primary asset being acted upon.[/italic dim]")
            elif func == "Detect":
                console.print(f"{tip_prefix} [italic dim]Map based on the Use Case (e.g., Insider Threat maps to Users).[/italic dim]")
            elif func == "Respond":
                console.print(f"{tip_prefix} [italic dim]Map based on the asset being investigated or contained.[/italic dim]")

            while True:
                try:
                    tech = questionary.text(
                        f"    Technology/Vendor:",
                        instruction=tech_instruction,
                        qmark=""
                    ).unsafe_ask()
                    data[asset][func]["Tech"] = tech if tech else ""

                    people = questionary.text(
                        f"    People/Responsible Role:",
                        instruction=people_instruction,
                        qmark=""
                    ).unsafe_ask()
                    data[asset][func]["People"] = people if people else ""

                    process = questionary.text(
                        f"    Process/Procedure:",
                        instruction=process_instruction,
                        qmark=""
                    ).unsafe_ask()
                    data[asset][func]["Process"] = process if process else ""
                    break

                except (KeyboardInterrupt, EOFError):
                    console.print("")
                    if questionary.confirm("Are you sure you want to quit?", qmark="", default=False).ask():
                        save_choice = questionary.select(
                            "Save progress before exiting?",
                            choices=["Save as Excel", "Save as JSON", "Quit without saving"],
                            qmark="",
                        ).ask()
                        if save_choice == "Save as Excel":
                            return data
                        elif save_choice == "Save as JSON":
                            json_path = questionary.text("JSON output path:", default="cdm_progress.json", qmark="").ask()
                            save_to_json(data, json_path)
                            console.print(f"[green]Progress saved to {json_path}[/green]")
                            sys.exit(0)
                        else:
                            console.print("[red]Quitting without saving.[/red]")
                            sys.exit(0)
                    console.print("[yellow]Resuming current function...[/yellow]")

    return data

def display_summary(data):
    """
    Displays a summary table of the collected data in the terminal, followed by a gap analysis.
    """
    function_list = functions()
    table = Table(title="Cyber Defense Matrix Summary", show_lines=True, expand=True)

    table.add_column("Asset", justify="left", style="bold cyan", no_wrap=True, min_width=12)
    for func in function_list:
        table.add_column(func, style="magenta")

    for asset, mappings in data.items():
        row_cells = [asset]
        for func in function_list:
            cell_data = mappings.get(func, {"Tech": "", "People": "", "Process": ""})
            tech    = cell_data["Tech"]    or "[dim]—[/dim]"
            people  = cell_data["People"]  or "[dim]—[/dim]"
            process = cell_data["Process"] or "[dim]—[/dim]"
            summary = f"[bold]Tech:[/bold] {tech}\n[bold]People:[/bold] {people}\n[bold]Process:[/bold] {process}"
            row_cells.append(summary)
        table.add_row(*row_cells)

    console.print(table)

    # Gap analysis
    full_count = partial_count = empty_count = 0
    empty_cells = []
    partial_cells = []

    for asset, mappings in data.items():
        for func in function_list:
            cell = mappings.get(func, {})
            filled = sum(1 for f in DATA_FIELDS if cell.get(f, "").strip())
            total_fields = len(DATA_FIELDS)
            if filled == total_fields:
                full_count += 1
            elif filled == 0:
                empty_count += 1
                empty_cells.append(f"{asset}/{func}")
            else:
                partial_count += 1
                partial_cells.append(f"{asset}/{func} ({filled}/{total_fields} fields)")

    total_cells = len(data) * len(function_list)
    status_color = "green" if empty_count == 0 else ("yellow" if empty_count <= 5 else "red")

    gap_lines = [
        f"[{status_color}][bold]{full_count}/{total_cells} cells fully populated[/bold][/{status_color}]"
        f"  •  [yellow]{partial_count} partial[/yellow]"
        f"  •  [red]{empty_count} empty[/red]"
    ]

    if empty_cells:
        gap_lines.append("\n[bold red]Empty cells (no data entered):[/bold red]")
        for name in empty_cells:
            gap_lines.append(f"  [red]✗[/red] {name}")

    if partial_cells:
        gap_lines.append("\n[bold yellow]Partial cells (some fields missing):[/bold yellow]")
        for name in partial_cells:
            gap_lines.append(f"  [yellow]△[/yellow] {name}")

    console.print(Panel("\n".join(gap_lines), title="[bold]Gap Analysis[/bold]", border_style="blue"))

File: gosrc/internal/wizard/wizard.go
package wizard

import (
	"fmt"
	"os"
	"strings"

	"cdmbuddy/internal/model"
	"github.com/charmbracelet/huh"
	"github.com/charmbracelet/lipgloss"
)

var (
	styleTitle   = lipgloss.NewStyle().Bold(true).Foreground(lipgloss.Color("12")).MarginBottom(1)
	styleHeader  = lipgloss.NewStyle().Bold(true).Foreground(lipgloss.Color("14")).Border(lipgloss.NormalBorder(), false, false, true, false).BorderForeground(lipgloss.Color("14"))
	styleIcon    = lipgloss.NewStyle().MarginRight(1)
	styleDim     = lipgloss.NewStyle().Italic(true).Faint(true)
	styleYellow  = lipgloss.NewStyle().Foreground(lipgloss.Color("11"))
	styleBlue    = lipgloss.NewStyle().Foreground(lipgloss.Color("12"))
	styleRed     = lipgloss.NewStyle().Foreground(lipgloss.Color("9"))
	styleReverse = lipgloss.NewStyle().Reverse(true).Bold(true).Padding(0, 1)
)

func RunWizard() (model.Matrix, error) {
	matrix := model.EmptyMatrix()

	// 1. Intro
	intro := huh.NewNote().
		Title("Cyber Defense Matrix Wizard").
		Description("Guided by Sounil Yu's 'The Essential Guide to Navigating the Cybersecurity Landscape'\n\nThe CDM is a MECE framework designed to provide structural clarity and identify gaps in your security posture.")

	concepts := huh.NewNote().
		Title("Core Concepts").
		Description(
			"• Left of Boom (Proactive): Identify and Protect. Focuses on Structural Awareness.\n" +
				"• Right of Boom (Reactive): Detect, Respond, and Recover. Focuses on Situational Awareness.\n\n" +
				"The Dependency Continuum:\n" +
				"• Technology dependency is strongest in Identify/Protect.\n" +
				"• People dependency grows significantly in Detect/Respond/Recover.\n" +
				"• Process is the essential, consistent foundation throughout all functions.",
		)

	form := huh.NewForm(huh.NewGroup(intro, concepts))
	if err := form.Run(); err != nil {
		return nil, err
	}

	// 2. Iterative Wizard
	orderedFunctions := []string{}
	for _, f := range model.Functions {
		if !contains(model.GovernFunctions, f) {
			orderedFunctions = append(orderedFunctions, f)
		}
	}
	for _, f := range model.Functions {
		if contains(model.GovernFunctions, f) {
			orderedFunctions = append(orderedFunctions, f)
		}
	}

	totalCells := len(model.AssetClasses) * len(orderedFunctions)
	cellNum := 0

	for assetIdx, asset := range model.AssetClasses {
		for funcIdx, funcName := range orderedFunctions {
			cellNum++
			progress := fmt.Sprintf("(Asset %d/%d | Function %d/%d | Cell %d/%d)", assetIdx+1, len(model.AssetClasses), funcIdx+1, len(orderedFunctions), cellNum, totalCells)

			var boomTag string
			var peopleInstruction string
			var processInstruction string

			if contains(model.GovernFunctions, funcName) {
				boomTag = styleYellow.Render("Cross-cutting")
				peopleInstruction = "e.g. Security Governance Lead, Risk Committee"
				processInstruction = "e.g. Risk Exception Process, Security Policy Review"
			} else {
				isRight := contains(model.RightOfBoomFunctions, funcName)
				if isRight {
					boomTag = styleRed.Render("Right of Boom")
				} else {
					boomTag = styleBlue.Render("Left of Boom")
				}
				peopleInstruction = "e.g. SOC Analyst, SysAdmin"
				processInstruction = "e.g. Patch Management SOP, IR Plan"
			}

			techInstruction := model.TechExamples[asset+"-"+funcName]
			if techInstruction == "" {
				techInstruction = "e.g. vendor tool or platform name"
			}

			icon := model.AssetIcons[asset]
			assetHeader := styleHeader.Render(fmt.Sprintf("%s %s", icon, strings.ToUpper(asset)))
			assetDesc := styleDim.Render(model.AssetDescriptions[asset])
			funcHeader := styleReverse.Render(" " + funcName + " ")
			funcDesc := styleDim.Render(model.FunctionDescriptions[funcName])

			var tip string
			tipPrefix := styleYellow.Render("💡 Tip:")
			switch funcName {
			case "Govern":
				tip = styleDim.Render("Think of this as the 'Context'—policies, risk appetite, and oversight governing this asset class.")
			case "Identify", "Protect", "Recover":
				tip = styleDim.Render("Map based on the primary asset being acted upon.")
			case "Detect":
				tip = styleDim.Render("Map based on the Use Case (e.g., Insider Threat maps to Users).")
			case "Respond":
				tip = styleDim.Render("Map based on the asset being investigated or contained.")
			}

			cell := matrix[asset][funcName]
			
			// Creating a group for this specific cell
			group := huh.NewGroup(
				huh.NewNote().Description(fmt.Sprintf("%s\n%s\n\n%s %s %s\n%s\n%s %s", assetHeader, assetDesc, funcHeader, boomTag, styleDim.Render(progress), funcDesc, tipPrefix, tip)),
				huh.NewInput().
					Title("Technology/Vendor").
					Placeholder(techInstruction).
					Value(&cell.Tech),
				huh.NewInput().
					Title("People/Responsible Role").
					Placeholder(peopleInstruction).
					Value(&cell.People),
				huh.NewInput().
					Title("Process/Procedure").
					Placeholder(processInstruction).
					Value(&cell.Process),
			)

			form := huh.NewForm(group)
			err := form.Run()
			
			if err != nil {
				// Handle interrupt
				if err == huh.ErrUserAborted {
					var quit bool
					confirmForm := huh.NewForm(huh.NewGroup(
						huh.NewConfirm().
							Title("Are you sure you want to quit?").
							Value(&quit),
					))
					_ = confirmForm.Run()
					if quit {
						var saveChoice string
						saveForm := huh.NewForm(huh.NewGroup(
							huh.NewSelect[string]().
								Title("Save progress before exiting?").
								Options(
									huh.NewOption("Save as Excel", "excel"),
									huh.NewOption("Save as JSON", "json"),
									huh.NewOption("Quit without saving", "quit"),
								).
								Value(&saveChoice),
						))
						_ = saveForm.Run()
						if saveChoice == "excel" {
							return matrix, nil
						} else if saveChoice == "json" {
							var path string
							pathForm := huh.NewForm(huh.NewGroup(
								huh.NewInput().Title("JSON output path").Value(&path).Placeholder("cdm_progress.json"),
							))
							_ = pathForm.Run()
							if path == "" { path = "cdm_progress.json" }
							_ = model.SaveToJson(matrix, path)
							fmt.Printf("Progress saved to %s\n", path)
							os.Exit(0)
						} else {
							fmt.Println("Quitting without saving.")
							os.Exit(0)
						}
					}
					// If they didn't quit, we loop back to the same cell by decrementing or just letting it continue
					// But huh doesn't easily allow jumping back in a simple loop without re-running.
					// Let's just continue the loop which will effectively re-ask this cell since cell data is passed by pointer.
					// Actually we need to make sure 'cell' is updated in the matrix.
					cellNum-- 
					continue 
				}
				return nil, err
			}
			matrix[asset][funcName] = cell
		}
	}

	return matrix, nil
}

func contains(slice []string, val string) bool {
	for _, item := range slice {
		if item == val {
			return true
		}
	}
	return false
}

func DisplaySummary(matrix model.Matrix) {
	// Simple summary for now, we can enhance with a table library later if needed
	fmt.Println("\nCyber Defense Matrix Summary")
	fmt.Println(strings.Repeat("-", 80))
	
	for _, asset := range model.AssetClasses {
		fmt.Printf("Asset: %s\n", asset)
		for _, funcName := range model.Functions {
			cell := matrix[asset][funcName]
			if cell.Tech != "" || cell.People != "" || cell.Process != "" {
				fmt.Printf("  [%s] Tech: %s, People: %s, Process: %s\n", funcName, cell.Tech, cell.People, cell.Process)
			}
		}
	}
	
	// Gap Analysis
	fullCount := 0
	partialCount := 0
	emptyCount := 0
	var emptyCells []string
	var partialCells []string

	for _, asset := range model.AssetClasses {
		for _, funcName := range model.Functions {
			cell := matrix[asset][funcName]
			filled := 0
			if cell.Tech != "" { filled++ }
			if cell.People != "" { filled++ }
			if cell.Process != "" { filled++ }

			if filled == 3 {
				fullCount++
			} else if filled == 0 {
				emptyCount++
				emptyCells = append(emptyCells, fmt.Sprintf("%s/%s", asset, funcName))
			} else {
				partialCount++
				partialCells = append(partialCells, fmt.Sprintf("%s/%s (%d/3 fields)", asset, funcName, filled))
			}
		}
	}

	totalCells := len(model.AssetClasses) * len(model.Functions)
	fmt.Printf("\nGap Analysis: %d/%d cells fully populated. %d partial, %d empty.\n", fullCount, totalCells, partialCount, emptyCount)
	
	if len(emptyCells) > 0 {
		fmt.Println("Empty cells:")
		for _, c := range emptyCells {
			fmt.Printf("  - %s\n", c)
		}
	}
}
